Windows BSP: `request failed: buildTarget/dependencySources` - cannot import into IDE - forward slash is prepended to the paths for some reason

Hi, I installed mill via scoop on my Windows 10 machine, JDK 16; `build.sc` is as follows:

```
import mill._
import scalalib._
object foo extends ScalaModule {
  def scalaVersion = "2.13.5"
}
```

Building with mill works; then I run `mill mill.bsp.BSP/install`, and import the project via IntelliJ's BSP support.

After import, it shows this:
![image](https://user-images.githubusercontent.com/2464813/115154149-4ee04000-a071-11eb-99d4-845d6e0c40af.png)

I am very keen on mill however last time I tried it it also had some other loading issue (pre-BSP times though iirc).

Stack trace:

```
Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Users/Dev/AppData/Local/Coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/mill-scalalib_2.13/0.9.6/mill-scalalib_2.13-0.9.6-sources.jar
    at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
    at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
    at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
    at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
    at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:230)
    at java.base/java.nio.file.Path.of(Path.java:147)
    at java.base/java.nio.file.Paths.get(Paths.java:69)
    at os.PathConvertible$StringConvertible$.apply(Path.scala:484)
    at os.PathConvertible$StringConvertible$.apply(Path.scala:483)
    at os.Path$.apply(Path.scala:382)
    at mill.bsp.ModuleUtils$.$anonfun$getMillBuildClasspath$10(ModuleUtils.scala:131)
    at mill.bsp.ModuleUtils$.$anonfun$getMillBuildClasspath$10$adapted(ModuleUtils.scala:131)
    at scala.collection.StrictOptimizedIterableOps.filterImpl(StrictOptimizedIterableOps.scala:226)
    at scala.collection.StrictOptimizedIterableOps.filterImpl$(StrictOptimizedIterableOps.scala:221)
    at scala.collection.immutable.ArraySeq.filterImpl(ArraySeq.scala:35)
    at scala.collection.StrictOptimizedIterableOps.filter(StrictOptimizedIterableOps.scala:217)
    at scala.collection.StrictOptimizedIterableOps.filter$(StrictOptimizedIterableOps.scala:217)
    at scala.collection.immutable.ArraySeq.filter(ArraySeq.scala:35)
    at mill.bsp.ModuleUtils$.getMillBuildClasspath(ModuleUtils.scala:131)
    at mill.bsp.MillBuildServer.$anonfun$buildTargetDependencySources$2(MillBuildServer.scala:136)
    at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:638)
    at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:634)
    at scala.collection.AbstractIterable.foldLeft(Iterable.scala:920)
    at mill.bsp.MillBuildServer.$anonfun$buildTargetDependencySources$1(MillBuildServer.scala:134)
    at mill.bsp.MillBuildServer.handleExceptions(MillBuildServer.scala:495)
    at mill.bsp.MillBuildServer.buildTargetDependencySources(MillBuildServer.scala:129)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:567)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
    ... 10 more



Apr 18, 2021 6:04:00 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError

SEVERE: Internal error: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Program%20Files/OpenJDK/src.zip

java.util.concurrent.CompletionException: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Program%20Files/OpenJDK/src.zip
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
    at java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:743)
    at java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731)
    at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2178)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:279)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
    at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consume(TracingMessageConsumer.java:91)
    at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
    at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
    at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
    at java.base/java.lang.Thread.run(Thread.java:831)

Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Program%20Files/OpenJDK/src.zip
    at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
    at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
    at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
    at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
    at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:230)
    at java.base/java.nio.file.Path.of(Path.java:147)
    at java.base/java.nio.file.Paths.get(Paths.java:69)
    at os.PathConvertible$StringConvertible$.apply(Path.scala:484)
    at os.PathConvertible$StringConvertible$.apply(Path.scala:483)
    at os.Path$.apply(Path.scala:382)
    at mill.bsp.ModuleUtils$.$anonfun$getMillBuildClasspath$10(ModuleUtils.scala:131)
    at mill.bsp.ModuleUtils$.$anonfun$getMillBuildClasspath$10$adapted(ModuleUtils.scala:131)
    at scala.collection.StrictOptimizedIterableOps.filterImpl(StrictOptimizedIterableOps.scala:226)
    at scala.collection.StrictOptimizedIterableOps.filterImpl$(StrictOptimizedIterableOps.scala:221)
    at scala.collection.immutable.ArraySeq.filterImpl(ArraySeq.scala:35)
    at scala.collection.StrictOptimizedIterableOps.filter(StrictOptimizedIterableOps.scala:217)
    at scala.collection.StrictOptimizedIterableOps.filter$(StrictOptimizedIterableOps.scala:217)
    at scala.collection.immutable.ArraySeq.filter(ArraySeq.scala:35)
    at mill.bsp.ModuleUtils$.getMillBuildClasspath(ModuleUtils.scala:131)
    at mill.bsp.MillBuildServer.$anonfun$buildTargetScalacOptions$2(MillBuildServer.scala:394)
    at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:638)
    at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:634)
    at scala.collection.AbstractIterable.foldLeft(Iterable.scala:920)
    at mill.bsp.MillBuildServer.$anonfun$buildTargetScalacOptions$1(MillBuildServer.scala:391)
    at mill.bsp.MillBuildServer.handleExceptions(MillBuildServer.scala:495)
    at mill.bsp.MillBuildServer.buildTargetScalacOptions(MillBuildServer.scala:386)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:567)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
    ... 10 more

```

--- a/bsp/src/mill/bsp/ModuleUtils.scala
+++ b/bsp/src/mill/bsp/ModuleUtils.scala
@@ -11,7 +11,11 @@ import mill.define._
 import mill.eval.{Evaluator, _}
 import mill.scalajslib.ScalaJSModule
 import mill.scalalib._
-import mill.scalalib.Lib.{depToDependency, resolveDependencies, scalaRuntimeIvyDeps}
+import mill.scalalib.Lib.{
+  depToDependency,
+  resolveDependencies,
+  scalaRuntimeIvyDeps
+}
 import mill.scalalib.api.Util
 import mill.scalanativelib._
 import mill.util.Ctx
@@ -36,10 +40,12 @@ object ModuleUtils {
   /**
    * Resolve a mill modules given a target identifier
    * */
-  def getModule(targetId: BuildTargetIdentifier, modules: Seq[JavaModule]): JavaModule =
+  def getModule(targetId: BuildTargetIdentifier,
+                modules: Seq[JavaModule]): JavaModule =
     modules
       .find(getTargetId(_) == targetId)
-      .getOrElse(throw new IllegalArgumentException(s"No module found for target id ${targetId.getUri}"))
+      .getOrElse(throw new IllegalArgumentException(
+        s"No module found for target id ${targetId.getUri}"))
 
   /**
    * Compute mapping between all the JavaModules contained in the
@@ -47,13 +53,14 @@ object ModuleUtils {
    * BSP BuildTargets ( mill modules correspond one-to-one to
    * bsp build targets ).
    *
-    * @param modules            All JavaModules contained in the working
+   * @param modules            All JavaModules contained in the working
    *                           directory of the mill project
    * @param evaluator          The mill evaluator that can resolve information
    *                           about the mill project
    * @return JavaModule -> BuildTarget mapping
    */
-  def getTargets(modules: Seq[JavaModule], evaluator: Evaluator)(implicit ctx: Ctx.Log): Seq[BuildTarget] = {
+  def getTargets(modules: Seq[JavaModule], evaluator: Evaluator)(
+      implicit ctx: Ctx.Log): Seq[BuildTarget] = {
     val targets = modules.map(module => getTarget(module, evaluator))
     val millBuildTarget = getMillBuildTarget(evaluator, modules)
 
@@ -61,7 +68,8 @@ object ModuleUtils {
   }
 
   def getMillBuildTargetId(evaluator: Evaluator): BuildTargetIdentifier =
-    new BuildTargetIdentifier(evaluator.rootModule.millSourcePath.toNIO.toUri.toString)
+    new BuildTargetIdentifier(
+      evaluator.rootModule.millSourcePath.toNIO.toUri.toString)
 
   /**
    * Compute the BuildTarget for the Mill build (build.sc files)
@@ -70,7 +78,8 @@ object ModuleUtils {
    *                    build information
    * @return the Mill BuildTarget
    */
-  def getMillBuildTarget(evaluator: Evaluator, modules: Seq[JavaModule])(implicit ctx: Ctx.Log): BuildTarget = {
+  def getMillBuildTarget(evaluator: Evaluator, modules: Seq[JavaModule])(
+      implicit ctx: Ctx.Log): BuildTarget = {
     val target = new BuildTarget(
       getMillBuildTargetId(evaluator),
       Seq.empty[String].asJava,
@@ -78,13 +87,16 @@ object ModuleUtils {
       Seq.empty[BuildTargetIdentifier].asJava,
       new BuildTargetCapabilities(false, false, false)
     )
-    target.setBaseDirectory(evaluator.rootModule.millSourcePath.toNIO.toUri.toString)
+    target.setBaseDirectory(
+      evaluator.rootModule.millSourcePath.toNIO.toUri.toString)
     target.setDataKind(BuildTargetDataKind.SCALA)
-    target.setTags(Seq(BuildTargetTag.LIBRARY, BuildTargetTag.APPLICATION).asJava)
+    target.setTags(
+      Seq(BuildTargetTag.LIBRARY, BuildTargetTag.APPLICATION).asJava)
     target.setDisplayName("mill-build")
 
     val scalaOrganization = "org.scala-lang"
-    val scalaLibDep = scalaRuntimeIvyDeps(scalaOrganization, BuildInfo.scalaVersion)
+    val scalaLibDep =
+      scalaRuntimeIvyDeps(scalaOrganization, BuildInfo.scalaVersion)
 
     val repos = Evaluator
       .evalOrElse(evaluator, T.traverse(modules)(_.repositoriesTask), Seq.empty)
@@ -111,24 +123,46 @@ object ModuleUtils {
     target
   }
 
-  def getMillBuildClasspath(evaluator: Evaluator, sources: Boolean): Seq[String] = {
-    val classpath = Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader]).fold(
-      _ => Seq.empty,
-      _.allJars
-    )
+  def getMillBuildClasspath(evaluator: Evaluator,
+                            sources: Boolean): Seq[String] = {
 
-    val millJars = resolveDependencies(
+    /** On Windows, URLs follow an peculiar representation in Java
+     * scala> java.nio.file.Paths.get(".").toAbsolutePath.toUri.toURL
+     * java.net.URL = file:/C:/Users/Developer/mill/./
+     *
+     * From this, we wish to get a Path back, and the way to do this is:
+     *
+     * scala> java.nio.file.Paths.get((java.nio.file.Paths.get(".").toAbsolutePath.toUri.toURL).toURI)
+     * java.nio.file.Path = C:\Users\Developer\mill\.
+     *
+     * Unit testing for this is more challenging because the WindowsFileSystem instance is a sun.nio.fs package,
+     * rather than a standard package.
+     *
+     * The solution here, compared to the previous code, is to reduce the number of conversions;
+     * the key loss happens when you do (URL).getFile.
+     * scala> java.nio.file.Paths.get(".").toAbsolutePath.toUri.toURL
+     * java.net.URL = file:/C:/Users/Developer/mill/./
+     * scala> java.nio.file.Paths.get(".").toAbsolutePath.toUri.toURL.getFile
+     * String = /C:/Users/Developer/mill/./
+     *  */
+    val classpath: Seq[Path] = Try(
+      evaluator.rootModule.getClass.getClassLoader
+        .asInstanceOf[SpecialClassLoader])
+      .fold(_ => Seq.empty, _.allJars)
+      .map(url => Path(java.nio.file.Paths.get(url.toURI)))
+
+    val millJars: Seq[Path] = resolveDependencies(
       Resolve.defaultRepositories,
       depToDependency(_, BuildInfo.scalaVersion),
       BuildInfo.millEmbeddedDeps.map(d => ivy"$d"),
       sources = sources
-    ).asSuccess.toSeq.flatMap(_.value).map(_.path.toNIO.toUri.toURL)
+    ).asSuccess.toSeq.flatMap(_.value).map(_.path)
 
     val all = classpath ++ millJars
     val binarySource =
-      if (sources) all.filter(url => isSourceJar(url))
-      else all.filter(url => !isSourceJar(url))
-    binarySource.filter(url => exists(Path(url.getFile))).map(_.toURI.toString)
+      if (sources) all.filter(url => isPathSourceJar(url))
+      else all.filter(url => !isPathSourceJar(url))
+    binarySource.filter(path => exists(path)).map(_.wrapped.toString)
   }
 
   /**
@@ -136,7 +170,7 @@ object ModuleUtils {
    * JavaModule, which is any module present in the working
    * directory, but it's not the root module itself.
    *
-    * @param module      any in-project mill module
+   * @param module      any in-project mill module
    * @param evaluator   mill evaluator
    * @return inner BuildTarget
    */
@@ -145,14 +179,18 @@ object ModuleUtils {
     val capabilities = getModuleCapabilities(module)
     val buildTargetTag = module match {
       case _: TestModule => Seq(BuildTargetTag.TEST)
-      case _: JavaModule => Seq(BuildTargetTag.LIBRARY, BuildTargetTag.APPLICATION)
+      case _: JavaModule =>
+        Seq(BuildTargetTag.LIBRARY, BuildTargetTag.APPLICATION)
     }
 
     val buildTarget = new BuildTarget(
       getTargetId(module),
       buildTargetTag.asJava,
       Seq("scala", "java").asJava,
-      (module.moduleDeps ++ module.compileModuleDeps).map(getTargetId).toList.asJava,
+      (module.moduleDeps ++ module.compileModuleDeps)
+        .map(getTargetId)
+        .toList
+        .asJava,
       capabilities
     )
     if (module.isInstanceOf[ScalaModule]) {
@@ -164,10 +202,11 @@ object ModuleUtils {
   }
 
   // obtain the capabilities of the given module ( ex: canCompile, canRun, canTest )
-  private[this] def getModuleCapabilities(module: JavaModule): BuildTargetCapabilities = {
+  private[this] def getModuleCapabilities(
+      module: JavaModule): BuildTargetCapabilities = {
     val canTest = module match {
       case _: TestModule => true
-      case _ => false
+      case _             => false
     }
 
     new BuildTargetCapabilities(true, canTest, true)
@@ -177,7 +216,7 @@ object ModuleUtils {
    * Evaluate the given task using the given mill evaluator and return
    * its result of type Result
    *
-    * @param evaluator mill evalautor
+   * @param evaluator mill evalautor
    * @param task      task to evaluate
    * @tparam T
    */
@@ -189,16 +228,18 @@ object ModuleUtils {
    * Evaluate the given task using the given mill evaluator and return
    * its result of type T, or the default value of the evaluation failed.
    *
-    * @param evaluator    mill evalautor
+   * @param evaluator    mill evalautor
    * @param task         task to evaluate
    * @param defaultValue default value to return in case of failure
    * @tparam T
    */
-  def evaluateInformativeTask[T](evaluator: Evaluator, task: Task[T], defaultValue: T): T = {
+  def evaluateInformativeTask[T](evaluator: Evaluator,
+                                 task: Task[T],
+                                 defaultValue: T): T = {
     val evaluated = evaluator.evaluate(Strict.Agg(task)).results(task)
     evaluated match {
       case Success(_) => evaluated.asSuccess.get.value.asInstanceOf[T]
-      case _ => defaultValue
+      case _          => defaultValue
     }
   }
 
@@ -207,19 +248,28 @@ object ModuleUtils {
       (module.millOuterCtx.millSourcePath / module.millModuleSegments.parts).toNIO.toUri.toString
     )
 
-  def getTarget(moduleHashCode: Int, modules: Seq[JavaModule], evaluator: Evaluator): Option[BuildTarget] =
+  def getTarget(moduleHashCode: Int,
+                modules: Seq[JavaModule],
+                evaluator: Evaluator): Option[BuildTarget] =
     modules.find(_.hashCode == moduleHashCode).map(getTarget(_, evaluator))
 
-  def getTargetId(moduleHashCode: Int, modules: Seq[JavaModule]): Option[BuildTargetIdentifier] =
+  def getTargetId(moduleHashCode: Int,
+                  modules: Seq[JavaModule]): Option[BuildTargetIdentifier] =
     modules.find(_.hashCode == moduleHashCode).map(getTargetId)
 
   def isSourceJar(url: URL): Boolean = url.getFile.endsWith("-sources.jar")
 
+  def isPathSourceJar(path: Path): Boolean =
+    path.wrapped.toString.endsWith("-sources.jar")
+
   // Compute the ScalaBuildTarget from information about the given JavaModule.
-  private[this] def computeBuildTargetData(module: JavaModule, evaluator: Evaluator): ScalaBuildTarget = {
+  private[this] def computeBuildTargetData(
+      module: JavaModule,
+      evaluator: Evaluator): ScalaBuildTarget = {
     module match {
       case m: ScalaModule =>
-        val scalaVersion = evaluateInformativeTask(evaluator, m.scalaVersion, "")
+        val scalaVersion =
+          evaluateInformativeTask(evaluator, m.scalaVersion, "")
         new ScalaBuildTarget(
           evaluateInformativeTask(evaluator, m.scalaOrganization, ""),
           scalaVersion,
@@ -240,28 +290,41 @@ object ModuleUtils {
           Seq.empty[String].asJava
         )
       case m =>
-        throw new IllegalStateException(s"Module type of ${m.millModuleSegments.render} not supported by BSP")
+        throw new IllegalStateException(
+          s"Module type of ${m.millModuleSegments.render} not supported by BSP")
     }
   }
 
   // Compute all relevant scala dependencies of `module`, like scala-library, scala-compiler,
   // and scala-reflect
-  private[this] def computeScalaLangDependencies(module: ScalaModule, evaluator: Evaluator): Agg[PathRef] = {
-    evaluateInformativeTask(evaluator, module.resolveDeps(module.scalaLibraryIvyDeps), Agg.empty[PathRef]) ++
-      evaluateInformativeTask(evaluator, module.scalacPluginClasspath, Agg.empty[PathRef]) ++
-      evaluateInformativeTask(evaluator, module.resolveDeps(module.ivyDeps), Agg.empty[PathRef]).filter(pathRef =>
-        pathRef.path.toNIO.toUri.toString.contains("scala-compiler") ||
-          pathRef.path.toNIO.toUri.toString.contains("scala-reflect") ||
-          pathRef.path.toNIO.toUri.toString.contains("scala-library")
-      )
+  private[this] def computeScalaLangDependencies(
+      module: ScalaModule,
+      evaluator: Evaluator): Agg[PathRef] = {
+    evaluateInformativeTask(
+      evaluator,
+      module.resolveDeps(module.scalaLibraryIvyDeps),
+      Agg.empty[PathRef]) ++
+      evaluateInformativeTask(
+        evaluator,
+        module.scalacPluginClasspath,
+        Agg.empty[PathRef]) ++
+      evaluateInformativeTask(
+        evaluator,
+        module.resolveDeps(module.ivyDeps),
+        Agg.empty[PathRef]).filter(
+        pathRef =>
+          pathRef.path.toNIO.toUri.toString.contains("scala-compiler") ||
+            pathRef.path.toNIO.toUri.toString.contains("scala-reflect") ||
+            pathRef.path.toNIO.toUri.toString.contains("scala-library"))
   }
 
   // Obtain the scala platform for `module`
-  private[this] def getScalaTargetPlatform(module: ScalaModule): ScalaPlatform = {
+  private[this] def getScalaTargetPlatform(
+      module: ScalaModule): ScalaPlatform = {
     module match {
       case _: ScalaNativeModule => ScalaPlatform.NATIVE
-      case _: ScalaJSModule => ScalaPlatform.JS
-      case _: ScalaModule => ScalaPlatform.JVM
+      case _: ScalaJSModule     => ScalaPlatform.JS
+      case _: ScalaModule       => ScalaPlatform.JVM
     }
   }
 }

